code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/PushTemplateImageUtil.kt

```diff
@@ -42,105 +43,158 @@ internal object PushTemplateImageUtil {
     private const val FULL_BITMAP_QUALITY = 100
     private const val DOWNLOAD_TIMEOUT_SECS = 10
 
-    private val executor: ExecutorService by lazy { Executors.newSingleThreadExecutor() }
-
-    private class DownloadImageCallable(val url: String?) :
-        Callable<Bitmap?> {
-        override fun call(): Bitmap? {
-            var bitmap: Bitmap? = null
-            var connection: HttpURLConnection? = null
-            try {
-                val imageUrl = URL(url)
-                connection = imageUrl.openConnection() as HttpURLConnection
-                connection.inputStream.use { inputStream ->
-                    bitmap = BitmapFactory.decodeStream(inputStream)
+    /**
+     * Downloads an image using the provided url `String`. Prior to downloading, the image url
+     * is used to retrieve a [CacheResult] containing a previously cached image. If no cache
+     * result is returned, a call to [download] is made to download then cache the image.
+     *
+     * If a valid cache result is returned then no image is downloaded.
+     *
+     * @param cacheService the AEPSDK [CacheService] to use for caching downloaded image assets
+     * @param urlList [String] containing an image asset url
+     * @return [Int] number of images that were found in cache or successfully downloaded
+     */
+    internal fun downloadImage(
+        cacheService: CacheService,
+        urlList: List<String?>
+    ): Int {
+        val assetCacheLocation = getAssetCacheLocation()
+        if (urlList.isEmpty() || assetCacheLocation.isNullOrEmpty()) {
+            return 0
+        }
+
+        var downloadedImageCount = 0
+        val latch = CountDownLatch(urlList.size)
+        for (url in urlList) {
+            if (url == null || !UrlUtils.isValidUrl(url)) {
+                latch.countDown()
+                continue
+            }
+
+            val cacheResult = cacheService[assetCacheLocation, url]
+            if (cacheResult != null) {
+                Log.trace(
+                    PushTemplateConstants.LOG_TAG,
+                    SELF_TAG,
+                    "Found cached image for $url"
+                )
+                downloadedImageCount++
+                latch.countDown()
+                continue
+            }
+
+            download(url) { image ->
+                image?.let {
                     Log.trace(
                         PushTemplateConstants.LOG_TAG,
                         SELF_TAG,
-                        "Downloaded push notification image from url ($url)"
+                        "Successfully download image from $url"
                     )
+                    downloadedImageCount++
+
```

Comment:
Careful here! `downloadedImageCount` is being modified by callbacks invoked from different threads spun by the NetworkService. So there is a possibility of a bad write on this variable where  two or more threads could read the value of `downloadedImageCount`, increment it, and write it back at the same time. This could result in `downloadedImageCount` being incremented only once instead of multiple times, leading to a lesser value than the actual number of images downloaded. This will result in an unnecessary wait even when all images have successfully been downloaded.

Can you make `downloadedImageCount` and `AtomicInteger` and `incrementAndGet` for modifying its value.

Reply:
Forgot to add this in a synchronized block, changing it